src/main/java/com/google/sps/servlets/BlobServlet.java

```diff
@@ -33,59 +41,203 @@
+    Key assetParentKey = projKey;
+    Entity imgEntity = new Entity("Image", projKey);
 
-    // Get the name entered by the user
-    String name = request.getParameter("name");
+    // Must be an owner or editor
+    Query projQuery = new Query("Project");
+
+    Filter ownEditFilter = new CompositeFilter(
+        CompositeFilterOperator.AND,
+        Arrays.asList(
+            new FilterPredicate("proj-id", FilterOperator.EQUAL, projId),
+            new CompositeFilter(
+                CompositeFilterOperator.OR,
+                Arrays.<Filter>asList(
+                    new FilterPredicate("owners", FilterOperator.EQUAL, uid),
+                    new FilterPredicate("editors", FilterOperator.EQUAL,
+                                        uid)))));
+
+    projQuery.setFilter(ownEditFilter);
+    PreparedQuery accessibleProjects = datastore.prepare(projQuery);
+
+    // Only owners and editors of given project can modify or create
+    if (accessibleProjects.countEntities() == 0) {
+      throw new IOException(
+          "User does not have permission to edit the project.");
+    }
+
+    // Asset to update must already exist
+    if (isMask) {
+      Query imgQuery = new Query("Image").setAncestor(projKey);
+      Filter imgFilter =
+          new FilterPredicate("name", FilterOperator.EQUAL, parentImg);
+      imgQuery.setFilter(imgFilter);
+      PreparedQuery existingImg = datastore.prepare(imgQuery);
+
+      if (existingImg.countEntities() == 0) {
+        throw new IOException(
+            "No parent image with that name exists for this project or this project does not exist.");
+      }
+      if (mode == "create") {
+        Key parentEntityKey =
+            existingImg.asSingleEntity().getKey();
+        imgEntity = new Entity("Mask", parentEntityKey);
+      } else {
+        assetParentKey = existingImg.asSingleEntity().getKey();
+
+        Query maskQuery = new Query("Mask").setAncestor(assetParentKey);
+        Filter maskFilter =
+            new FilterPredicate("name", FilterOperator.EQUAL, imgName);
+        imgQuery.setFilter(maskFilter);
+        PreparedQuery existingMask = datastore.prepare(maskQuery);
+
+        if (existingMask.countEntities() == 0) {
+          throw new IOException(
+              "No mask with that name exists to update for this project.");
+        }
+        imgEntity = existingMask.asSingleEntity();
+      }
+    } else if (mode == "update") {
+      Query imgQuery = new Query("Image").setAncestor(projKey);
+      Filter imgFilter =
+          new FilterPredicate("name", FilterOperator.EQUAL, imgName);
+      imgQuery.setFilter(imgFilter);
+      PreparedQuery existingImg = datastore.prepare(imgQuery);
+
+      if (existingImg.countEntities() == 0) {
+        throw new IOException(
+            "No image with that name exists to update for this project or this project does not exist.");
+      }
+      imgEntity = existingImg.asSingleEntity();
+    }
```

Comment:
This set of logic is difficult to understand the intention.  I am wondering why you switch on IsMask and mode == Update.  Why not if on only mode (create/update)?  Why don't we handle not mask and create?

Reply:
isMask && create is the default mode which is why no specific block addresses it. I will refactor this logic and add more comments.